src/ShellClients/ShellClientsManager.vala

```diff
@@ -166,37 +166,45 @@ public class Gala.ShellClientsManager : Object {
      * TODO: Maybe use for strut only?
      */
     public void set_size (Meta.Window window, int width, int height) {
-        if (!(window in windows)) {
+        if (!(window in panel_windows)) {
             warning ("Set anchor for window before size.");
             return;
         }
 
-        windows[window].set_size (width, height);
+        panel_windows[window].set_size (width, height);
     }
 
     public void set_hide_mode (Meta.Window window, Pantheon.Desktop.HideMode hide_mode) {
-        if (!(window in windows)) {
+        if (!(window in panel_windows)) {
             warning ("Set anchor for window before hide mode.");
             return;
         }
 
-        windows[window].set_hide_mode (hide_mode);
+        panel_windows[window].set_hide_mode (hide_mode);
     }
 
-    public void make_centered (Meta.Window window) {
-        if (window in centered_windows) {
-            return;
-        }
+    public void make_centered (Meta.Window window) requires (!is_itself_positioned (window)) {
+        positioned_windows[window] = new WindowPositioner (window, wm, (ref x, ref y) => {
```

Comment:
I wonder is having a whole position func here kind of overkill? Do we need things other than like corners and center?

Reply:
Corners, center, panel (so top and bottom) and maybe monitor corner without the padding of the monitor labels if we want to go with something like #2070.
I've got no strong preference here, we can move all types of "positioning funcs" into the positioner. It's not quite that straight forward because e.g. the monitor labels need additional info (the monitor index) but that can be done quite easily with a Variant or something (to keep it flexibel for the future :)).

Reply:
I've moved the positioning logic to the positioner now to distribute the code more evenly :sweat_smile: